AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/SubscriptionSync/ReconcileAndLocalSave/ReconcileAndLocalSaveQueue.swift

```diff
@@ -0,0 +1,116 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Amplify
+import Combine
+import Foundation
+import AWSPluginsCore
+
+@available(iOS 13.0, *)
+protocol ReconcileAndSaveOperationQueue {
+    func addOperation(_ operation: ReconcileAndLocalSaveOperation, modelName: String)
+
+    func cancelOperations(modelName: String)
+
+    func cancelAllOperations()
+
+    func waitUntilOperationsAreFinished()
+
+    init(_ modelSchemas: [ModelSchema])
+}
+
+enum ReoncileAndSaveQueueEvent {
+    case operationAdded(id: UUID)
+    case operationRemoved(id: UUID)
+    case cancelledOperations(modelName: String)
+}
+
+/// A queue used to enqueue `ReconcileAndLocalSaveOperation`s which perform reconcile and save for incoming mutation
+/// sync events to local storage for all model types.
+///
+/// Internally, a `ReconcileAndSaveQueue` will manage an operation queue with concurrency of 1 to perform serial
+/// operations to ensure events are processed in the order in which they are received. For example, an initial sync
+/// will perform sync queries for models based on its dependency order. The results are then processed serially by this
+/// queue to ensure the events are reconciled in the same dependency order in which they are received by the
```

Comment:
It seems misleading to tie our reconciliation to the order in which we receive events. Isn't it accurate to say that the reconciliation order is determined by our inspection of the dependency graph? That is, in a graph where **A** depends on **B** depends on **C**, as long as we reconcile all of **C** first, then all of **B**, then all of **A**, the order in which we receive any individual **C** event is irrelevant.

Is my thinking right here?

Reply:
yes that's right, the other callout would be regarding subscription events which come in the correct order of indivudal events B then A close to the same time. In this case we should be performing the reconcilliation in the order which any events are received to ensure we don't process A before B. 
```
/// Internally, a `ReconcileAndSaveQueue` will manage an operation queue with concurrency of 1 to perform serial
/// operations to ensure models are processed in the order of dependency. For example, an initial sync
/// will perform sync queries for models based on its dependency order. The results are then processed serially by this
/// queue to ensure that a model A that depends on B, and B depends on C, will be reconciled in the order of C then B
/// then A. This also ensures that reconciliation for individual subscription events are also processed in the order
/// which they are received by the system.
```